blocks/log.py

```diff
@@ -1,26 +1,318 @@
 """The event-based main loop of Blocks."""
-from collections import defaultdict
+import sqlite3
+from abc import ABCMeta
+from collections import defaultdict, MutableMapping, Mapping
+from itertools import repeat
 from numbers import Integral
+from operator import itemgetter
+from uuid import uuid4, UUID
 
+from six import add_metaclass
+from six.moves import cPickle, map
 
-class TrainingLog(defaultdict):
-    """Base class for training logs.
+from blocks.config import config
+
+
+def _sub_string(replacements):
+    """Return a certain number of ? to be substituted by `replacements`."""
+    return '({})'.format(', '.join(repeat('?', len(replacements))))
+
+
+@add_metaclass(ABCMeta)
+class _TrainingLog(object):
+    """Base class for training log.
 
     A training log stores the training timeline, statistics and other
     auxiliary information. Information is stored as a nested dictionary,
     ``log[time][key]``. An entry without stored data will return an empty
-    dictionary (like ``defaultdict(dict)``).
+    dictionary-like object that can be written to (like
+    ``defaultdict(dict)``). However, entries need to be added one by one,
```

Comment:
Yeah. But I just realised that `log[t].update({'foo': 'bar'})` should work for all backends, so I'll recommend that.